### PR TITLE
Fix covariant returns with generic return types

### LINKED ISSUE #45037: ExecutionEngineException in assembly.GetTypes() when the assembly has a class with covariant return types
### Description
Whenever I try to run `assembly.GetTypes()` on an assembly with a class similar to the one below I get an ExecutionEngineException. A minimal reproduction can be found here: https://github.com/roy-t/ExecutionEngineExceptionReproduction. 

```
public abstract class ABase
{
    public abstract object this[int index] { get; }
}

public sealed class Concrete<T> : ABase
    where T : class
{
    public override T this[int index] // <- covariant indexer
    {
       get
       {
            throw null;
        }
    }
}
```

I think the problem is the covariant indexer. Whenever I replace the covariant indexer with a regular one that just returns `object` the problem goes away. I've also tried to reproduce the problem with methods and properties with covariant return types, but that did not trigger the bug.

**Update**: after more experimentation I did find it possible to trigger this bug with covariant return types for methods and properties. See my comment down below.

### Configuration
I'm running .NET 5.0  (v5.0.100) on Window 10 x64. I have not tried it on another machine yet.

### Regression?

Not really, since .NET 5.0 is the first version that supports covariant return types. However in previous .net versions you were able to call `GetTypes()` on an assembly without a crash :).

### Other information
Unfortunately there's no stack trace. The only information I get from Visual Studio is:

```
An unhandled exception of type 'System.ExecutionEngineException' occurred in System.Private.CoreLib.dll

The program '[17496] FooCrash.exe' has exited with code -1 (0xffffffff).
```

By the way, I hope this is the right GitHub project for these types of exceptions. I saw that System.Private.Corelib lives in this repository so it was my best guess :)

### LINKED ISSUE #45082: Mixing covariant return types with generics triggers a TypeLoadException in some cases
### Description
After playing with covariant return types and reporting #45037 I think I found another problem with covariant return types. I've written a piece of code that compiles without problems, and to me logically makes sense. But as soon as the class with the method that has a covariant return type is used I get a TypeLoadException. I've managed to bring down the reproduction to running the following code.

```
public abstract class AComponent { }
public class Component : AComponent { }

public abstract class Abstract
{
    public abstract IReadOnlyList<AComponent> New { get; }
}

public sealed class Concrete<T> : Abstract
    where T : AComponent
{
    public override IReadOnlyList<T> New => throw null;
}

internal class Program
{
    private static void Main(string[] args)
    {
        new Concrete<Component>(); 
    }
}
```
Simplifying this even further, like replacing `AComponent` and `Component` with `object` and `string` no longer triggers the bug, which really puzzles me. You also need some generic to put the type that is actually 'varying' into. I've chosen `IReadonlyList<>` here. But using another interface also works as long as it has an `out` modifier for the type parameter. (Otherwise the code won't compile, which makes sense). 

### Configuration

I'm running .NET 5.0 (v5.0.100) on Window 10 x64. I have not tried it on another machine yet.

### Regression?

Not really since covariant return types are a new feature in .NET 5.0.

### Other information
The complete error message is:

```
System.TypeLoadException
  HResult=0x80131522
  Message=Return type in method 'MakeGenericBug.Concrete`1[System.__Canon].get_New()' on type 'MakeGenericBug.Concrete`1[System.__Canon]' from assembly 'MakeGenericBug, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not compatible with base type method 'MakeGenericBug.Abstract.get_New()'.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>
```

Unfortunately there's no stack trace, I think the exception happens before encountering my code since no matter where I set a break-point, its never hit.

### PR DESCRIPTION
**Pull request by janvorli**
When a covariant return type was an uninstantiated generic type, the
ClassLoader::IsCompatibleWith was not working properly. In debug builds,
it was asserting because there was no MethodTable for that type and in
release builds, it resulted in ExecutionEngineException or an internal
CLR error.

This change fixes it by using TypeHandle::CanCastTo instead of
MethodTable::CanCastTo and adds a regression test for two cases where
the problem was observed (Assembly.GetTypes() and creating an instance
of a type with a covariant return with a problematic kind of type).

Close #45037
Close #45082

### CODE REVIEW

**Comment by benaadams** [comment]
Also fixes https://github.com/dotnet/runtime/issues/45082 ?

**Comment by janvorli** [comment]
@benaadams I've just tried and it doesn't fix it. There is yet another issue hidden behind my fix (with my fix, it fails at a different place than before).

**Comment by benaadams** [comment]
I did try changing the runtime to use covariant returns for RuntimeType; but it then crashes with both types of crossgen and jitdiff https://github.com/benaadams/runtime/commit/e19bcfd7a3af1f76273051bcc8dd1579be40446a

**Comment by janvorli** [comment]
@benaadams can you please create an issue for that? I'll look into it after fixing the #45082.

**Comment by benaadams** [comment]
Raised https://github.com/dotnet/runtime/issues/45356

**Comment by janvorli** [comment]
I've updated the PR so that it fixes the #45082 too and added a regression test for that one.

File: src/coreclr/src/vm/clsload.cpp [hunk 1]
```diff
@@ -3531,16 +3531,15 @@ static void PushFinalLevels(TypeHandle typeHnd, ClassLoadLevel targetLevel, cons
...
-        }
-
         DFLPendingList pendingList;
         BOOL           fBailed = FALSE;
 
         typeHnd.DoFullyLoad(NULL, CLASS_LOADED, &pendingList, &fBailed, pInstContext);
 
+        if (!typeHnd.IsTypeDesc())
+        {
+            ClassLoader::ValidateMethodsWithCovariantReturnTypes(typeHnd.AsMethodTable());
```
Comment (jkotas) [inline src/coreclr/src/vm/clsload.cpp:3541]: DoFullyLoad marks the type as fully loaded. Another thread that tries to load the same type may see it as fully loaded and won't perform this validation.

I think this needs to be done before the type is marked as fully loaded.

Comment (janvorli) [reply src/coreclr/src/vm/clsload.cpp:3541]: Hmm, so I'll need to figure out how to fix the issue that was fixed by this change. The problem was that down the call chain of the ValidateMethodsWithCovariantReturnTypes, a load level assert fired.

Comment (davidwrighton) [reply src/coreclr/src/vm/clsload.cpp:3541]: This is the only concern I have with this change.

**Review by davidwrighton — CHANGES_REQUESTED** [review]
Please fix the issue around DoFullyLoad

**Review by davidwrighton — APPROVED** [review]

**Comment by janvorli** [comment]
The `Libraries Test Run release mono windows x64 Debug` leg keeps failing consistently in many libraries tests (even after closing / reopening the issue) and I cannot find previous failures of this leg. Moreover, this change doesn't touch anything from Mono stuff, so I am baffled what's going on. @jkotas or @marek-safar do you have any idea what can be going on?

**Comment by jkotas** [comment]
@akoeplinger @safern Could you please disable the failing Mono legs or move them to the staging pipeline so that they do not cause all jobs to be red?

**Comment by akoeplinger** [comment]
The leg was already disabled with https://github.com/dotnet/runtime/pull/45529 but you'd need to rebase/merge master to get this change in the PR. I'd just merge.

**Comment by janvorli** [comment]
Thanks @akoeplinger!

**Pull request merged**